contracts/zero-ex/tests/Multiplex.t.sol

```diff
@@ -0,0 +1,651 @@
+            rfqOrder.takerToken,
+            zrx,
+            _makeArray(
+                _makeRfqSubcall(rfqOrder),
+                _makeUniswapV2BatchSubcall(_makeArray(address(dai), address(zrx)), rfqOrder.takerAmount, false)
+            ),
+            rfqOrder.takerAmount,
+            0
+        );
+    }
+
+    // expired OTC, fallback(UniswapV2)
+    function test_multiplexBatchSellTokenForToken_expiredOtcFallbackUniswapV2() public {
+        LibNativeOrder.OtcOrder memory otcOrder = _makeTestOtcOrder();
+        address uniV2Pool = _createUniswapV2Pool(uniV2Factory, dai, zrx, 10e18, 10e18);
+        _mintTo(address(otcOrder.takerToken), otcOrder.taker, otcOrder.takerAmount);
+        otcOrder.expiryAndNonce = 1;
+
+        vm.expectEmit(true, true, true, true);
+        emit ExpiredOtcOrder(
+            zeroExDeployed.features.otcOrdersFeature.getOtcOrderHash(otcOrder),
+            otcOrder.maker,
+            uint64(otcOrder.expiryAndNonce >> 192)
+        );
+
+        vm.expectEmit(true, true, true, true);
+        emit Transfer(address(dai), otcOrder.taker, uniV2Pool, otcOrder.takerAmount);
+
+        vm.expectEmit(true, true, true, true);
+        emit Transfer(address(zrx), uniV2Pool, otcOrder.taker, 906610893880149131);
+
+        zeroExDeployed.zeroEx.multiplexBatchSellTokenForToken(
+            otcOrder.takerToken,
+            zrx,
+            _makeArray(
+                _makeOtcSubcall(otcOrder),
+                _makeUniswapV2BatchSubcall(_makeArray(address(dai), address(zrx)), otcOrder.takerAmount, false)
+            ),
+            otcOrder.takerAmount,
+            0
+        );
+    }
+
+    // expired RFQ, fallback(TransformERC20)
+    function test_multiplexBatchSellTokenForToken_expiredRfqFallbackTransformErc20() public {
+        LibNativeOrder.RfqOrder memory rfqOrder = _makeTestRfqOrder();
+        _mintTo(address(rfqOrder.takerToken), rfqOrder.taker, rfqOrder.takerAmount);
+        rfqOrder.expiry = 0;
+
+        vm.expectEmit(true, true, true, true);
+        emit ExpiredRfqOrder(
+            zeroExDeployed.features.nativeOrdersFeature.getRfqOrderHash(rfqOrder),
+            rfqOrder.maker,
+            rfqOrder.expiry
+        );
+
+        vm.expectEmit(true, true, true, true);
+        emit Transfer(address(dai), rfqOrder.taker, address(flashWallet), rfqOrder.takerAmount);
+
+        vm.expectEmit(true, true, true, true);
+        emit MintTransform(
+            address(flashWallet),
+            address(zeroExDeployed.zeroEx),
+            address(zeroExDeployed.zeroEx),
+            rfqOrder.taker,
+            abi.encode(dai, zrx, rfqOrder.takerAmount, 5e17, 0),
+            rfqOrder.takerAmount,
+            0
+        );
+
+        vm.expectEmit(true, true, true, true);
+        emit Transfer(address(dai), address(flashWallet), address(0), 1e18);
+
+        vm.expectEmit(true, true, true, true);
+        emit Transfer(address(zrx), address(flashWallet), rfqOrder.taker, 5e17);
```

Comment:
It looks like you would rather just check the events vs check the token balances for these tests. Wouldn't checking the balances be a more reliable way of verifying the trade went through?

Reply:
I think there's some value in checking both. Making the tests robust so that if something strange changes the test starts failing is a good idea.

For example, the way these `vm.expectEmit`s are structure, if the exchange proxy emits the `Transfer` event (or the wrong token, or any other contract involved), the test will still pass. Checking the balances in addition to the events would catch that.

FWIW, there's a `vm.expectEmit(bool, bool, bool, bool, address)` variant that would fix :point_up: